### ISSUE TITLE
Improve the implementation of dask.array.block

### ISSUE THREAD

**Issue by hmaarrfk** [dask/dask#3987]
https://github.com/dask/dask/blob/14b4c3848516a0f03bcf1b40e3f1f7475131497c/dask/array/core.py#L2899

I recently rewrote what will likely be numpy's new implementation of `block`. @eric-wieser suggested an enhancement which should make it really trivial for dask to piggy back off the results and use it in it's implementation.

https://github.com/numpy/numpy/pull/11971#issuecomment-421910996

The first time I tried to use dask.block, it was really slow, so I had to write my own. I think this might help it since it avoid calls to concatenate all together.

Not sure about unknown dimensions.

**Comment by mrocklin** [comment]
Improvements are always welcome.  Two questions:

1.  That implementation seems to use `__setitem__` syntax, which probably won't be as useful for dask array.  Does that affect your plans?
2.  Why was dask.array.block slow?  Was it slow in graph construction overhead, or in execution?

**Comment by hmaarrfk** [comment]
1. Eventually we will have a helper function that returns the `dtype`, `shape`, `slices` and `numpy arrays`. These could easily be `dask arrays`. An other function will take all of those and create a new array and do the assignment. I Think that is within the realm of dask since assignment only happens to newly created arrays.

2. Not too sure. It probably wasn't any slower than block. I think ultimately it failed at handling one of my cases, so I couldn't use it. 

I just wanted to point somebody to a `no concatenate` solution to this problem.

**Comment by mrocklin** [comment]
This might also be helpful in replacing the current `concatenate3` function in dask.array.core, which also does a no-concatenate blocking solution, but is probably uglier than what numpy has.  If numpy.block is now convenient and efficient then it might be good to switch over, especially if there is a performance increase.

**Comment by mrocklin** [comment]
Any update on this @hmaarrfk ?

**Comment by hmaarrfk** [comment]
I started with a benchmark. It shows basically that it suffers a little bit from overhead. But not from copying data.

**Comment by jakirkham** [comment]
FWIW a lot of the code for `block` was lifted from NumPy. This was done to avoid NumPy usages of `asarray`, which trigger a computation, or things like `concatenate`, where a Dask friendly implementation is needed. If there have been changes in NumPy, that would help Dask, it is probably reasonable to update our vendored copy. Better yet if we can consolidate code between the two, but that likely isn't possible without NumPy addressing `asarray` and `concatenate` for this use case in a nicer way.

**Comment by mrocklin** [comment]
Checking in here.  Any update on this @hmaarrfk ?  Is there anything that you are blocked on here on the dask side?

**Comment by hmaarrfk** [comment]
No, more I'm waiting for the numpy PR to go through. A few peculiarities have come up.

**Comment by mrocklin** [comment]
OK, just checking in.  Thanks for the update.

**Comment by hmaarrfk** [comment]
Yeah, so while the dask operation is built on multiple dask implementations
of concatenate, those aren't actually costly as data isn't copied until the
very end.

Maybe graph operations can be reduced, making the graph cleaner, but it
isn't something I am enthousiastic about because that numpy PR took so long
to get through, probably in large part due to my novice development style.

I am secretly hoping now that numpy is expanding their high level
interface, that maybe we can expose the internal algorithm of block and use
that for dask. Dask would depend on numpy 1.16 or 1.17

I'm not sure what would happen when you block two dask arrays with chunks
that don't line up. An operation that is probably allowed now.
